File: ros2_ws/src/arena_camera_node/arena_camera_node/trigger_image_client.py
import rclpy
#from arena_camera_software_trigger.srv import TriggerImage
from rclpy.node import Node
from std_srvs.srv import Trigger


class TriggerClientNode(Node):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        log_debug = self.get_logger().debug
        log_info = self.get_logger().info

        self.cli = self.create_client(Trigger, 'trigger_image')

        while not self.cli.wait_for_service(timeout_sec=0.5):
            log_info(f'service not available, waiting again... ->')
        log_debug('service is available now')
        self.req = Trigger.Request()
        log_debug('trigger request object has been created')

    def send_request(self):
        self.future = self.cli.call_async(self.req)


def main(args=None):

    log_debug = None
    log_info = None

    rclpy.init(args=args)

    trigger_client = TriggerClientNode(node_name='trigger_image_client_node')
    log_debug = trigger_client.get_logger().debug
    log_info = trigger_client.get_logger().info
    log_debug('trigger_image_client_node has been created')

    trigger_client.send_request()
    log_info('trigger_image_client_node sent a request')

    while rclpy.ok():
        log_debug('trigger_image_client is waiting for a response')

        rclpy.spin_once(trigger_client)
        log_debug('trigger_image_client spinned once')

        log_debug('checking if responce is received?')
        if trigger_client.future.done():
            log_debug('responce is received')
            try:
                response = trigger_client.future.result()
                log_debug('response is read')
                if not response.success:
                    log_info('response.success == False')
                    raise Exception(f'Failed to publish an image on trigger')
            except Exception as e:
                log_info('Service call failed %r' % (e,))
            else:
                log_info(response.message)
            log_debug('breaking the loop')
            break

    log_debug('trigger_image_client is destroyed')
    trigger_client.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_ws/src/arena_camera_node/src/trigger_image_client.cpp


// import rclpy
#include "rclcpp/rclcpp.hpp"

// from std_srvs.srv import Trigger
#include <chrono>
#include "rclcpp/client.hpp"
#include "std_srvs/srv/trigger.hpp"

using namespace std::chrono_literals;

class TriggerArenaImageClientNode : public rclcpp::Node
{
 public:
  explicit TriggerArenaImageClientNode(
      const std::string& node_name = "trigger_arena_image_client")
      : Node(node_name)
  {
    // TODO
    // - more generic name to all the trigger to work for many devices
    std::string srvs_name = "/arena_camera_node/trigger_image";

    m_cli_ = this->create_client<std_srvs::srv::Trigger>(srvs_name);

    // wait if the service is not available
    while (!m_cli_->wait_for_service(1s) && rclcpp::ok()) {
      if (!rclcpp::ok()) {
        log_err("Interrupted while waiting for the service. Exiting.");
        rclcpp::shutdown();
      } else {
        log_info(srvs_name + " service not available, waiting again... ->");
      }
    };
    // log_debug_('service is available now')
    // log_debug('trigger request object has been created')
  }
  void send_request()
  {
    auto req_ = std::make_shared<std_srvs::srv::Trigger::Request>();
    m_result = m_cli_->async_send_request(req_);
  }

  std::shared_future<std::shared_ptr<std_srvs::srv::Trigger_Response>> m_result;
  void log_info(const std::string& msg)
  {
    RCLCPP_INFO(this->get_logger(), msg);
  };
  void log_err(const std::string& msg)
  {
    RCLCPP_ERROR(this->get_logger(), msg);
  };

 private:
  rclcpp::Client<std_srvs::srv::Trigger>::SharedPtr m_cli_;
};

int main(int argc, char** argv)
{
  // TODO
  // - should take number of images to trigger
  // - add debug logs

  // force flush of the stdout buffer.
  // this ensures a correct sync of all prints
  // even when executed simultaneously within the launch file.
  setvbuf(stdout, NULL, _IONBF, BUFSIZ);

  rclcpp::init(argc, argv);

  auto client_node = std::make_shared<TriggerArenaImageClientNode>(
      "trigger_arena_image_client");
  client_node->send_request();
  // Wait for the result.
  if (rclcpp::spin_until_future_complete(client_node, client_node->m_result) ==
      rclcpp::executor::FutureReturnCode::SUCCESS) {
    auto response = client_node->m_result.get();
    if (response->success) {
      client_node->log_info(std::string("SUCCESS : ") + response->message);
    } else {
      client_node->log_err(std::string("FAILURE : ") + response->message);
    }
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"),
                 "Failed to trigger image; issue with spin");
  }

  rclcpp::shutdown();

  return 0;
}